### ISSUE TITLE
The four `declaration-block-semicolon-*` rules read a Sass nested property as a declaration closed by a semicolon, and it is closed by its block

### ISSUE THREAD

**Issue by firefoxic** [stylelint-stylistic/stylelint-stylistic#437]
Found while closing [#139](https://github.com/stylelint-stylistic/stylelint-stylistic/issues/139), which measures the same declarations and leaves this alone. Every warning below stands on `main` as it stands on that branch, and every fix output is byte for byte the same on both; the three positions #139 moves are named where they come up.

A Sass nested property is a declaration carrying a block, and nothing closes it but that block: `postcss-scss` reads `font: 12px` and a block of its own for it, and Sass compiles a file that writes no semicolon behind that block exactly as it compiles one that does. `isLastDeclarationWithoutSemicolon` passes such a declaration over only where it stands last in its block, so wherever anything stands behind it these four rules take it for a declaration carrying a semicolon and go looking for the whitespace around one.

### What minimal example or steps are needed to reproduce the bug?

```scss
a {
	font: 12px
	{ family: serif; }
	top: 0;
}
```

There is no semicolon anywhere on line 2 or line 3.

### What minimal configuration is needed to reproduce the bug?

```json
{
	"customSyntax": "postcss-scss",
	"plugins": ["@stylistic/stylelint-plugin"],
	"rules": {
		"@stylistic/declaration-block-semicolon-space-before": "never"
	}
}
```

### What did you expect to happen?

Nothing said about the nested property. The whitespace between its value and its block is that block's own indentation, and no option of these four rules is about it.

### What actually happened?

Over that file, every primary option of the four rules that says anything about the nested property:

| rule | option | warning | what `--fix` writes |
| --- | --- | --- | --- |
| `declaration-block-semicolon-space-before` | `always` | `3:1 Expected single space before ";"` | `font: 12px { family: serif; }` |
| `declaration-block-semicolon-space-before` | `never` | `3:1 Unexpected whitespace before ";"` | `font: 12px{ family: serif; }` |
| `declaration-block-semicolon-newline-before` | `never-multi-line` | `3:1 Unexpected whitespace before ";" in a multi-line declaration block` | `font: 12px{ family: serif; }` |
| `declaration-block-semicolon-space-after` | `always` | `4:1 Expected single space after ";"` | `{ family: serif; } top: 0;` |
| `declaration-block-semicolon-space-after` | `never` | `4:1 Unexpected whitespace after ";"` | `{ family: serif; }top: 0;` |
| `declaration-block-semicolon-newline-after` | `never-multi-line` | `4:1 Unexpected newline after ";" in a multi-line declaration block` | `{ family: serif; }top: 0;` |

The three positions reading `4:1` read `3:3` on `main`, inside the nested block — that column is #139's and moves with it. What the warnings say and what the fixes write is the same on both.

Sass compiles every one of those outputs to the same three declarations the file started with, so nothing is broken: what happens is that a warning is raised about a character the file does not hold, and `--fix` reformats the block's indentation, or the line break in front of the next declaration, against an option that was never about either. `declaration-block-semicolon-newline-before` under `always` and `always-multi-line` reports the declaration behind the nested property rather than the property itself, and is right about it.

### Do you have a proposal to fix the bug?

The question these rules ask is whether the declaration is closed by a semicolon at all, and `isLastDeclarationWithoutSemicolon` answers it for one place only — the end of a block. A declaration carrying a block of its own is closed by that block wherever it stands, so the guard wants widening rather than each of the four rules wanting a special case. Whether a semicolon written behind such a block belongs to the declaration or to the block around it is the same question `declaration-block-trailing-semicolon` meets in [#336](https://github.com/stylelint-stylistic/stylelint-stylistic/issues/336), and the two are worth answering together.

**Comment by firefoxic** [comment]
A measurement to add, found while reviewing the branch that closes #139.

Where the semicolon of a nested property is written against the block, the two `*-after` rules of the family are right about it; where anything stands in between, the column they report does not move at all. All three files below are the same three declarations to Sass.

```scss
a { font: 12px { family: serif; };top: 0; }
a { font: 12px { family: serif; }; top: 0; }
a { font: 12px { family: serif; } ; top: 0; }
```

Under `declaration-block-semicolon-space-after` with `always`, the first two are reported at `1:35`, and the semicolon stands at 34: that is the character behind it, which is what the rule reports for a plain declaration. The third is the one this is about. Under `never` it is reported at `1:35` as well, and there the closing brace stands at 33, the space at 34, the semicolon at 35 and the character behind it at 36. Widen that run and the reported column stays at 35 however wide it grows: `1:35` for one space, for two, for four, where the answer is 36, 37 and 39.

The reason it cannot move is that the run in front of such a semicolon is filed in the **next** declaration's `raws.before` — for the third file, `top`'s is `" ; "` — which is outside the node the rule measures, so the one character the rule steps over the semicolon with has nothing to step over that run with.

A plain declaration is not affected: PostCSS files that run inside the node, and `a { color: red ; top: 0; }` under `never` is reported at `1:17`, the character behind the semicolon, with the column tracking the run as it grows.

This is the same reading the issue is about, seen from the other side: the rule takes the semicolon to stand against the declaration because it takes the declaration to carry one. Whatever settles the guard settles this too.

On `main` all four of those warnings fall at `1:17`. For the three nested files that column is the space one past the nested block's opening brace, which is #139 rather than this, and the branch closing that issue moves them to the positions above. For the plain declaration `1:17` is already the right answer, and neither `main` nor that branch moves it.
